File: pegasusX/apps/supplier-app-android/app/src/main/java/com/pegasusx/supplier/ui/screens/catalog/CatalogBarcodeField.kt
package com.pegasusx.supplier.ui.screens.catalog

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import com.pegasus.barcode.EanBarcode
import com.pegasus.barcode.EanBarcodeScannerPreview
import com.pegasusx.supplier.ui.theme.PegasusSpacing
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

@Composable
fun CatalogBarcodeField(
    value: String,
    onValueChange: (String) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    var showScanner by rememberSaveable { mutableStateOf(false) }
    var scannerEnabled by rememberSaveable { mutableStateOf(true) }
    var validationError by rememberSaveable { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(PegasusSpacing.sm)) {
        OutlinedTextField(
            value = value,
            onValueChange = {
                validationError = null
                onValueChange(it)
            },
            label = { Text("EAN / GTIN barcode") },
            placeholder = { Text("8–14 digits or scan label") },
            keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number),
            singleLine = true,
            isError = validationError != null,
            supportingText = validationError?.let { { Text(it) } },
            modifier = Modifier.fillMaxWidth(),
            enabled = enabled,
        )
        TextButton(
            onClick = { showScanner = !showScanner },
            enabled = enabled,
        ) {
            Text(if (showScanner) "Hide camera" else "Scan barcode")
        }
        if (showScanner && enabled) {
            EanBarcodeScannerPreview(
                enabled = scannerEnabled,
                onBarcode = { scanned ->
                    val normalized = EanBarcode.normalize(scanned)
                    if (normalized == null) {
                        validationError = "Invalid EAN/GTIN — check the label"
                        return@EanBarcodeScannerPreview
                    }
                    validationError = null
                    onValueChange(normalized)
                    scannerEnabled = false
                    scope.launch {
                        delay(1500)
                        scannerEnabled = true
                    }
                },
            )
        }
        if (value.isNotBlank()) {
            val normalized = EanBarcode.normalize(value)
            Text(
                if (normalized != null) "Valid GTIN: $normalized" else "Enter a valid EAN/GTIN checksum",
                style = MaterialTheme.typography.labelSmall,
                color = if (normalized != null) {
                    MaterialTheme.colorScheme.tertiary
                } else {
                    MaterialTheme.colorScheme.error
                },
            )
        }
    }
}

File: pegasusX/apps/supplier-app-ios/SupplierApp/Views/Catalog/CatalogBarcodeField.swift
import SwiftUI

struct CatalogBarcodeField: View {
    @Binding var value: String
    var enabled: Bool = true

    @State private var showScanner = false
    @State private var scannerEnabled = true
    @State private var validationMessage: String?

    private var normalized: String? {
        EANBarcode.normalize(value)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: SupplierTheme.spacingSM) {
            TextField("mobile_supplier.ui.ean_gtin_barcode", text: $value)
                .keyboardType(.numberPad)
                .textInputAutocapitalization(.never)
                .autocorrectionDisabled()
                .disabled(!enabled)
                .onChange(of: value) { _, _ in
                    validationMessage = nil
                }

            Button(showScanner ? "Hide camera" : "Scan barcode") {
                showScanner.toggle()
            }
            .font(.caption)
            .disabled(!enabled)

            if showScanner && enabled {
                EANBarcodeScannerView(
                    onBarcode: { scanned in
                        guard let code = EANBarcode.normalize(scanned) else {
                            validationMessage = "Invalid EAN/GTIN — check the label"
                            return
                        }
                        validationMessage = nil
                        value = code
                        scannerEnabled = false
                        Task {
                            try? await Task.sleep(for: .milliseconds(1500))
                            scannerEnabled = true
                        }
                    },
                    enabled: scannerEnabled
                )
            }

            if let validationMessage {
                Text(validationMessage)
                    .font(.caption)
                    .foregroundStyle(.red)
            } else if !value.isEmpty {
                Text(normalized != nil ? "Valid GTIN: \(normalized!)" : "Enter a valid EAN/GTIN checksum")
                    .font(.caption)
                    .foregroundStyle(normalized != nil ? Color.secondary : Color.red)
            }
        }
    }
}
